freud/density.pyx

```diff
@@ -576,10 +576,17 @@ cdef class RDF(_SpatialHistogram1D):
     behavior of :math:`\lim_{r \to \infty} g(r)=1` is recovered. However, for
     very small systems the long range behavior of the radial distribution will
     instead tend to :math:`\frac{N-1}{N}`. In small systems, where this
-    deviation is noticeable, the ``normalize`` flag may be used to rescale the
-    results and force the long range behavior to 1. Note that this option will
-    have little to no effect on larger systems (for example, for systems of 100
-    particles the RDF will differ by 1%).
+    deviation is noticeable, the ``normalization_mode`` argument may be used to
+    rescale the results and force the long range behavior to 1. Note that this
+    option will have little to no effect on larger systems (for example, for
+    systems of 100 particles the RDF will differ by 1%).
+
+    .. note::
+        The ``normalization_mode`` argument should not be used if
+        :code:`query_points` is provided as a different set of points, or if
+        unusual query arguments are provided to :meth:`~.compute`, specifically
```

Comment:
unusual? You already explicitly mention which argument is problematic. Are there any others?

Reply:
Next sentence also states that if such unusual arguments are provided the data will be "convoluted"? Does that mean it will be wrong? I'd avoid using word convoluted here. In this specific context convolution can also be interpreted as a mathematical operation...

Reply:
I agree, this is poor wording. Let me try to explain it better.

If you follow through the derivation of the `n`-point correlation function for a system of `N` particles, you'll find that it has a prefactor of the form `N!/(N-n)!` For a 2-point correlation, this is simply `N * (N-1)`. There's an additional factor of `N^2` that pops up (IIRC from the volume integral, but it's been a long time so don't quote me on that) that results in a prefactor that looks like `(N-1)/N`. For an ideal gas, all the other terms in the correlation function are unity, so `g(r) -> (N-1)/N = 1 - 1/N`. In the thermodynamic limit of `N->\infty`, `1-1/N -> 1`.

The inconsistency that led to the addition of the `normalize` flag arises because when most packages compute the pair correlation function they end up just dividing by `N` in a place where freud chooses to be exact and divide by `N-1`. For large systems the difference is negligible, but it is noticeable when working with small systems (aka far from the thermodynamic limit). All that `normalize` does is multiply the result by `N/N-1` to flip between freud's convention and the convention used by other tools.

Unfortunately, the normalization by values like `N` is only applicable when working with a single set of particles. The actual values get more complicated when you want to compute, for example, the RDF between two different subsets of points in the system that represent two different types of particles. In that situation, the math demands a different normalization. Ultimately the normalization constant boils down to something proportional to the number of pairs of points considered for the RDF. Other tools like VMD or LAMMPS have access to the whole system, so when a user requests the RDF computation between two subsets of the system, those tools can explicitly compute the number of pairs by looking at how many distinct pairs exist between the two sets of particles. freud cannot do this because the only information freud has is the two sets of input positions; it doesn't know if some of those actually represent the same particle (and we don't want to try to infer that based on overlapping positions, that's out of scope for freud). 

The point of this warning is basically to state that the `N->N-1` transition is _only_ valid in a case where the RDF is computed based on a cutoff distance (`r_max`) where `exclude_ii=True` and `query_points=points`. If `exclude_ii=False`, then the division is already by `N` rather than `N-1` because we're including a particle's interaction with itself (this is where the `N-1` comes from in the normal case of `exclude_ii=True`) so the additional normalization would produce the wrong result. If `query_points != points`, then any simple rescaling will be wrong unless the two sets of points are completely disjoint (and freud can't know that), so the `normalize` behavior is pretty useless.

The purpose of switching from `normalize` to `normalize_mode` is to be able to handle some of the simplest cases where we know the normalization, even if it has a form different from just `N/N-1`. The obvious cases that we can handle are:
- `finite_size`: Apply the finite size correction of `N/N-1` so that even for small system sizes `g(r)->1` as `r->\infty`. This option will give a user incorrect results if they specify `exclude_ii=False` and will result in 
`g(r)->1+1/N`.
- `infer`: If `query_points == points`, apply the finite size correction, otherwise don't. This option will give incorrect results if, for instance, `query_points != points` but there are some shared points between the two sets.
- `none`: Don't apply any normalization at all. This mode doesn't currently exist, but would essentially not do any division based on `N` and leave that up to the user. This mode is what advanced users will need if they use `query_points` that are distinct from but have a nonempty intersection with `points`.

Hopefully that explanation clears up any confusion here.

Reply:
You're right, my original post was a little bit imprecise. What I was calling "infer" isn't exactly the current behavior. I had intended to use the current behavior when `query_points == points` and the finite size correct when `query_points == points`. That's why I called it "infer"; it infers the appropriate behavior in at least the two simplest cases.

I think we have two options:
1. Implement the "infer" behavior that I described above. It would be fairly easy, but it is nonzero work.
2. Just have the default remain exactly the same as the current behavior `normalization=False` but name the mode something (`exact`? not sure) to more precisely characterize what it's doing: it's the correct calculation when the RDF is computed for a single set of points against itself.

The more I think about it, the more inclined I am to choose option 2. RDF normalization is an esoteric enough topic that I would prefer not to have anything implicit.

Reply:
I think leaving the current behavior is fine, and calling it `exact` better reflects what its actually doing.